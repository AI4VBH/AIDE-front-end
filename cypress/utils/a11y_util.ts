import axe from "axe-core";
import { Options } from "cypress-axe";

export function nodeTerminal(violations: axe.Result[]) {
    cy.task(
        "log",
        `${violations.length} accessibility violation${violations.length === 1 ? "" : "s"} ${
            violations.length === 1 ? "was" : "were"
        } detected`,
    );
    // pluck specific keys to keep the table readable
    const violationData: Array<any> = violations.map(
        ({ id, impact, description, nodes, helpUrl }) => ({
            id,
            impact,
            description,
            nodes: nodes.length,
            helpUrl,
        }),
    );

    cy.task("table", violationData);
}

export const a11yConfig: Options = {
    runOnly: {
        type: "tags",
        values: ["wcag21a", "wcag21aa", "wcag2a", "wcag2aa"],
    },
};
