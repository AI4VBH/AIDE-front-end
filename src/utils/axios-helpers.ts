/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import axios, { AxiosError, AxiosInstance, AxiosRequestConfig, AxiosResponse } from "axios";
import Vue from "vue";

export interface ErrorMessageMap {
    get?: string;
    post?: string;
    put?: string;
    patch?: string;
    delete?: string;
}

export function isResultOk({ status }: AxiosResponse): boolean {
    return status >= 200 && status <= 299;
}

export function authenticationEnabledInterceptor(config: AxiosRequestConfig): AxiosRequestConfig {
    if (process.env.VUE_APP_AUTH_ENABLED === "true") {
        Vue.prototype.$keycloak?.keycloak.updateToken(70);
    }

    return config;
}

export function attachBearerTokenInterceptor(config: AxiosRequestConfig): AxiosRequestConfig {
    if (config.headers && Vue.prototype.$keycloak?.token) {
        config.headers["Authorization"] = `Bearer ${Vue.prototype.$keycloak.token}`;
    }

    return config;
}

export function onRequestErrorInterceptor(
    error: AxiosError,
    messageMap: ErrorMessageMap,
    conflictToastHidden?: boolean,
): Promise<AxiosError> {
    if (error.code === "ERR_NETWORK") {
        Vue.$toast?.error("Connection error. Please check your internet connection.");

        return Promise.reject(error);
    } else if (error.response?.status === 401) {
        Vue.prototype.$keycloak?.logout({ redirectUri: `${window.location.origin}/#/` });

        return Promise.reject(error);
    } else if (error.response?.status === 409 && conflictToastHidden) {
        return Promise.reject(error);
    }

    const messageKey = `${
        error.request?.method || error.config?.method
    }`.toLowerCase() as keyof ErrorMessageMap;
    const message = messageMap[messageKey] ?? "Something unexpected went wrong.";

    Vue.$toast?.error(message);

    return Promise.reject(error);
}

export function createAxiosInstance(
    errorMessages?: ErrorMessageMap,
    conflictToastHidden?: boolean,
): AxiosInstance {
    const http = axios.create({
        baseURL: window.FRONTEND_API_HOST,
        headers: {
            "Content-Type": "application/json",
        },
    });

    http.interceptors.request.use((config) => authenticationEnabledInterceptor(config));
    http.interceptors.request.use((config) => attachBearerTokenInterceptor(config));
    http.interceptors.response.use(
        (response) => response,
        (error) => onRequestErrorInterceptor(error, errorMessages ?? {}, conflictToastHidden),
    );

    return http;
}
