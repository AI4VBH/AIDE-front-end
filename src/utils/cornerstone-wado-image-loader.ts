/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import dicomParser from "dicom-parser";
import * as cornerstone from "@cornerstonejs/core";
import cornerstoneWADOImageLoader from "cornerstone-wado-image-loader";

export default function initCornerstoneWADOImageLoader(token?: string) {
    cornerstoneWADOImageLoader.external.cornerstone = cornerstone;
    cornerstoneWADOImageLoader.external.dicomParser = dicomParser;
    cornerstoneWADOImageLoader.configure({
        useWebWorkers: true,
        decodeConfig: {
            convertFloatPixelDataToInt: false,
        },
    });

    if (token) {
        cornerstoneWADOImageLoader.configure({
            beforeSend: (xhr: XMLHttpRequest) => {
                xhr.setRequestHeader("Authorization", `Bearer ${token}`);
            },
        });
    }

    let maxWebWorkers = 1;

    if (navigator.hardwareConcurrency) {
        maxWebWorkers = Math.min(navigator.hardwareConcurrency, 7);
    }

    const config = {
        maxWebWorkers,
        startWebWorkersOnDemand: false,
        taskConfiguration: {
            decodeTask: {
                initializeCodecsOnStartup: false,
                strict: false,
            },
        },
    };

    cornerstoneWADOImageLoader.webWorkerManager.initialize(config);
}
