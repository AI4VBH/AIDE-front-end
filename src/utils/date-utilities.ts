/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import moment from "moment";

// example:
// 27 Oct 2022 1.05pm
export const NhsDateTimeFormat = "DD-MMM-YYYY h:mma";

export const formatDateAndTimeOfArray = (
    items: any,
    itemProperty: string,
    inludeFromNow = true,
): void => {
    items.map((item: any) => {
        item[itemProperty] = formatDateAndTimeOfString(item[itemProperty], inludeFromNow);
    });
};

/// Returns date time for examples...
/// 27 Oct 2022 1.05pm (44 minutes ago)
/// 27 Oct 2022 1.05pm (20 years ago)
/// You can remove from now bit of sting by passning in false too inludeFromNow param.
export function formatDateAndTimeOfString(dateStr: string, inludeFromNow = true) {
    const dateMoment = moment(dateStr);
    if (!dateMoment.isValid()) {
        return "";
    }
    return inludeFromNow
        ? `${dateMoment.format(NhsDateTimeFormat)} (${dateMoment.fromNow()})`
        : `${dateMoment.format(NhsDateTimeFormat)}`;
}
