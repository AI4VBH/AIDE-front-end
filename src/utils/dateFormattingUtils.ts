export const formatDateAndTimeOfArray = (items: any, itemProperty: string): void[] => {
    return items.map((item: any) => {
        const date = item[itemProperty].split("T")[0].replace(/(\d{4})(\d{2})(\d+)/, "$1-$2-$3");
        const hour = Number(item[itemProperty].split("T")[1].substr(0, 2));
        const suffix: string = hour >= 12 ? "PM" : "AM";
        const formattedHour = (hour % 12 || 12).toString();
        const minutes = item[itemProperty].split("T")[1].substr(2, 2).toString();
        item[itemProperty] =
            date +
            " " +
            (Number(formattedHour) < 10 ? "0" + formattedHour : formattedHour) +
            ":" +
            minutes +
            " " +
            suffix;
    });
};

export const formatDateAndTimeOfString = (item: string): string => {
    const date = item.split("T")[0].replace(/(\d{4})(\d{2})(\d+)/, "$1-$2-$3");
    const hour = Number(item.split("T")[1].substr(0, 2));
    const minutes = item.split("T")[1].substr(2, 2).toString();
    return (item = date + " " + (Number(hour) < 10 ? "0" + hour : hour) + ":" + minutes);
};
