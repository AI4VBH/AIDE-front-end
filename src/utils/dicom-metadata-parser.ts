import parser, { DataSet } from "dicom-parser";
import tagLabels from "./dicom-tag-dictionary";

type DicomTag = {
    tag: string;
    vr: string;
    vm: string;
    name: string;
};

export type DicomMetadata = {
    name: string;
    value?: string;
};

function isAscii(text?: string): boolean {
    if (!text) {
        return false;
    }

    // eslint-disable-next-line no-control-regex
    return /^[\x00-\x7F]*$/.test(text);
}

function isStringVr(vr: string): boolean {
    const nonString = ["AT", "FL", "FD", "OB", "OF", "OW", "SI", "SQ", "SS", "UL", "US"];

    return !nonString.includes(vr);
}

function parseNumericalValues(
    metadataItem: DicomMetadata,
    propertyName: string,
    elementCount: number,
    elementDivider: number,
    extractor: (tag: string, index?: number) => number | undefined,
): DicomMetadata {
    metadataItem.value = `${extractor(propertyName)}`;

    for (let i = 1; i < elementCount / elementDivider; i++) {
        metadataItem.value += `\\${extractor(propertyName, i)}`;
    }

    return metadataItem;
}

function getDicomTag(tag: string): DicomTag {
    const tagIndex = `(${tag.substring(1, 5)},${tag.substring(5, 9)})`.toUpperCase();
    return tagLabels[tagIndex];
}

function dumpDataset(dataset: DataSet): DicomMetadata[] {
    const propertyNames = Object.keys(dataset.elements).sort();

    const metadata: DicomMetadata[] = [];

    propertyNames.forEach((propertyName) => {
        const element = dataset.elements[propertyName];

        if (element.tag <= "x0002ffff" || element.length <= 0 || parser.isPrivateTag(element.tag)) {
            return;
        }

        const tag = getDicomTag(element.tag);

        const metadataItem: DicomMetadata = {
            name: tag?.name ?? element.tag,
            value: undefined,
        };

        if (element.items) {
            const value: { [key: string]: string | undefined }[] = [];

            element.items.forEach((item) => {
                if (!item.dataSet) {
                    return;
                }

                const items = dumpDataset(item.dataSet);

                const vals = items.map(({ name, value }) => {
                    const obj: { [key: string]: string | undefined } = {};
                    obj[name] = value;

                    return obj;
                });

                value.push(...vals);
            });

            metadataItem.value = JSON.stringify(value, null, 2);
            metadata.push(metadataItem);

            return;
        } else if (element.fragments) {
            return;
        }

        if (!element.vr && !tag) {
            const data = dataset.string(propertyName);

            if (isAscii(data)) {
                metadataItem.value = data;
                metadata.push(metadataItem);

                return;
            }

            if (element.length !== 2 && element.length !== 4) {
                metadataItem.value = "binary data";
                metadata.push(metadataItem);

                return;
            }
        }

        const vr = element.vr ?? tag.vr;

        if (isStringVr(vr)) {
            const data = dataset.string(propertyName);

            if (isAscii(data)) {
                metadataItem.value = data;
                metadata.push(metadataItem);

                return;
            }

            if (element.length !== 2 && element.length !== 4) {
                metadataItem.value = "binary data";
                metadata.push(metadataItem);
            }
        } else if (vr === "US") {
            metadata.push(
                parseNumericalValues(metadataItem, propertyName, element.length, 2, (tag, index) =>
                    dataset.uint16(tag, index),
                ),
            );
        } else if (vr === "SS") {
            metadata.push(
                parseNumericalValues(metadataItem, propertyName, element.length, 2, (tag, index) =>
                    dataset.int16(tag, index),
                ),
            );
        } else if (vr === "UL") {
            metadata.push(
                parseNumericalValues(metadataItem, propertyName, element.length, 4, (tag, index) =>
                    dataset.uint32(tag, index),
                ),
            );
        } else if (vr === "SL") {
            metadata.push(
                parseNumericalValues(metadataItem, propertyName, element.length, 4, (tag, index) =>
                    dataset.int32(tag, index),
                ),
            );
        } else if (vr === "FD") {
            metadata.push(
                parseNumericalValues(metadataItem, propertyName, element.length, 8, (tag, index) =>
                    dataset.double(tag, index),
                ),
            );
        } else if (vr === "FL") {
            metadata.push(
                parseNumericalValues(metadataItem, propertyName, element.length, 4, (tag, index) =>
                    dataset.float(tag, index),
                ),
            );
        }
    });

    return metadata;
}

export function parseMetadata(byteArray: Uint8Array): DicomMetadata[] {
    const dataset = parser.parseDicom(byteArray, {
        vrCallback(tag) {
            const formatted = `(${tag.substring(1, 5)},${tag.substring(5, 9)})`.toUpperCase();
            return tagLabels[formatted] ? tagLabels[formatted].vr : undefined;
        },
    });

    return dumpDataset(dataset);
}

export function parseEncapsulatedPdf(byteArray: Uint8Array): Uint8Array | undefined {
    const dataset = parser.parseDicom(byteArray, {
        vrCallback(tag) {
            const formatted = `(${tag.substring(1, 5)},${tag.substring(5, 9)})`.toUpperCase();
            return tagLabels[formatted] ? tagLabels[formatted].vr : undefined;
        },
    });

    const documentData = dataset.elements["x00420011"];

    if (!documentData) {
        return;
    }

    return dataset.byteArray.slice(
        documentData.dataOffset,
        documentData.dataOffset + documentData.length,
    );
}
