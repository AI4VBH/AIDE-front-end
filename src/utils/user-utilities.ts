import Vue from "vue";

export enum UserRole {
    admin = "admin",
    clinician = "clinician",
    user_management = "user_management",
    deployer = "deployer",
}

export const getDefaultDestinationForUser = () => {
    if (process.env.VUE_APP_AUTH_ENABLED === "false") {
        return "AdminHealthDashboard";
    }

    const noRole = () => false;
    const hasRealmRole = Vue.prototype.$keycloak.hasRealmRole ?? noRole;

    if (hasRealmRole(UserRole.admin)) {
        return "AdminHealthDashboard";
    } else if (hasRealmRole(UserRole.clinician)) {
        return "ClinicalReview";
    } else if (hasRealmRole(UserRole.deployer)) {
        return "ApplicationRepositoryList";
    } else if (hasRealmRole(UserRole.user_management)) {
        return "UserManagement";
    }

    return "Unauthorized";
};
