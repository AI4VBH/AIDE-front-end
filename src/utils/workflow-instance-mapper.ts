/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { TaskExecution, WorkflowInstance } from "@/models/Admin/IPayload";

export interface ExecutionTreeFirstNode {
    id: string;
    name: string;
    workflow_name: string;
    workflow_instance_id: string;
    status: string;
    start_date: string;
    children: ExecutionTreeNode[];
}

export interface ExecutionTreeNode {
    id: string;
    execution_id: string;
    workflow_instance_id: string;
    name: string;
    status: string;
    start_date: string;
    children: ExecutionTreeNode[];
}

export interface ExecutionTreeRoot {
    id: "root-node";
    name: "Payload Received";
    status: string;
    children: ExecutionTreeFirstNode[];
}

export function mapToExecutionTreeFirstNode(instance: WorkflowInstance): ExecutionTreeFirstNode {
    return {
        id: "workflow-instance",
        name: instance.ae_title,
        workflow_name: instance.workflow_name,
        workflow_instance_id: instance.id,
        status: instance.status,
        start_date: instance.start_time,
        children: (instance.tasks ?? []).map((t) => mapToExecutionTreeNode(t)),
    };
}

export function mapToExecutionTreeNode(task: TaskExecution): ExecutionTreeNode {
    return {
        id: task.execution_id,
        execution_id: task.execution_id,
        workflow_instance_id: task.workflow_instance_id,
        name: task.task_id,
        status: task.status,
        start_date: task.task_start_time,
        children: (task.next_task ?? []).map((t) => mapToExecutionTreeNode(t)),
    };
}

export function mapToExecutionTree(instances: WorkflowInstance[]): ExecutionTreeRoot {
    return {
        id: "root-node",
        name: "Payload Received",
        status: "Succeeded",
        children:
            instances.length > 0
                ? instances.map((i) => mapToExecutionTreeFirstNode(i))
                : [mapToEmptyWorkflow()],
    };
}

export function mapToEmptyWorkflow(): ExecutionTreeFirstNode {
    return {
        id: "no-workflow-instance",
        name: "No Workflow Triggered",
        status: "Succeeded",
        workflow_name: "",
        workflow_instance_id: "",
        start_date: "",
        children: [],
    };
}
