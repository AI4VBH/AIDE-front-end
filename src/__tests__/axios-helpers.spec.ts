/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { isResultOk, provideDefaultResult } from "../utils/axios-helpers";
import { expect, test as it } from "@jest/globals";
import { AxiosResponse } from "axios";

describe("axios-helpers", () => {
    describe("isResultOk", () => {
        it.each([200, 255, 299])("returns true when status code is %s", (status) => {
            const result = isResultOk({ status } as AxiosResponse);
            expect(result).toBe(true);
        });

        it.each([100, 300, 404, 500])("returns false when status code is %s", (status) => {
            const result = isResultOk({ status } as AxiosResponse);
            expect(result).toBe(false);
        });
    });

    describe("provideDefaultResult", () => {
        it("returns expected result when status is OK", () => {
            const response = { status: 200, data: ["hello", "world"] } as AxiosResponse<string[]>;

            const result = provideDefaultResult(response, [":("]);

            expect(result).toMatchSnapshot();
        });

        it("returns expected result when status is not OK", () => {
            const response = { status: 400, data: ["hello", "world"] } as AxiosResponse<string[]>;

            const result = provideDefaultResult(response, [":("]);

            expect(result).toMatchSnapshot();
        });
    });
});
