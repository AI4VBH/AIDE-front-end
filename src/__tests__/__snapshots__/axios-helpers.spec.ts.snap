// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`axios-helpers provideDefaultResult returns expected result when status is OK 1`] = `
Array [
  "hello",
  "world",
]
`;

exports[`axios-helpers provideDefaultResult returns expected result when status is not OK 1`] = `
Array [
  ":(",
]
`;
